src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -281,7 +282,7 @@ public void stop(Throwable cause) throws Exception {
                             continue;
                         }
                         listener.getLogger().println("Agent " + node.getNodeName() + " was deleted; cancelling node body");
-                        task.body.cancel(new RemovedNodeCause());
+                        task.body.get().cancel(new RemovedNodeCause());
```

Comment:
Does the result of `WeakReference#get()` need to be surrounded with a `null` check? After reading the Javadoc more closely, I think it does. But I think I will wait for confirmation from a reviewer before I go through another testing cycle.

Reply:
I'm not sure, it might be non-null in this case but it's not obvious to me that the `WeakReference` could never be `null` here, so I would add `BodyExecution body = task.body == null ? null : task.body.get();`  on line 280 before the `task.body == null` check, and then change the check to `body == null`, and use `body` here at which point it would be guaranteed non-null.

Reply:
I'm not so sure either. `WeakReference#get` does state that the result can be "`null` if this reference object has been cleared". Presumably whatever is actually running the body has a reference to it until it stops running (which is why this implementation strategy is safe in the first place), but I'm not familiar enough with Pipeline internals to confirm this. In any case, I went ahead and added the `null` check as you suggested. I'm normally against overly-defensive `null` checks, but since I'm not the primary maintainer of this code it seemed prudent to avoid the risk of any potential regressions.